### PR TITLE
Search and Password reset

### PR DESCRIPTION
**Pull request by Ayoush**
Requirements

- 1. Feature to reset password
- 2. Add Indexing in Search functionality where ever we used elasticsearch is being used.

Added

1. Index functionality added to Widget Search
2. Index functionality added to User Search
3. Index functionality added to Asset Search
4. Test cases added to accommodate above changes.
5. Add search for sensors
6. Add search for Assets
7. Add search for Asset Types
8. Add search for Sensor Types

For forgot password:-
1. User will click forgot password link which will trigger one mail to
   user containing the reset link.
2. On clicking the link user will be redirected to a page where he will
   reset the password by entering the new one.

### CODE REVIEW

File: apps/acqdat_api/lib/acqdat_api_web/controllers/role-management/invitation_controller.ex [hunk 1]
```diff
@@ -28,6 +28,27 @@ defmodule AcqdatApiWeb.RoleManagement.InvitationController do
...
+
+          _token_details ->
+            conn
+            |> put_status(200)
+            |> json(%{is_valid: true})
+        end
+
+      404 ->
+        conn
+        |> send_error(404, "User already exists with this email address")
```
Comment (arjun289) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/role-management/invitation_controller.ex:48]: How are you coming to this conclusion. I couldn't find a reasonable logic to this?

File: apps/acqdat_api/test/acqdat_api_web/controllers/elasticsearch/asset_controller_test.exs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  alias AcqdatCore.Factory.ElasticSearch.Asset
+  import AcqdatCore.Support.Factory
+
+  describe "search_assets/2" do
+    setup :setup_conn
+
+    test "fails if authorization header not found", %{conn: conn} do
+      asset = insert(:asset)
+      Asset.seed_asset(asset)
+      :timer.sleep(2500)
```
Comment (arjun289) [inline apps/acqdat_api/test/acqdat_api_web/controllers/elasticsearch/asset_controller_test.exs:14]: No need to seed here as you are testing failure due to bad header.

File: apps/acqdat_api/test/acqdat_api_web/controllers/elasticsearch/asset_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    end
+  end
+
+  describe "index assets/2" do
+    setup :setup_conn
+
+    test "fails if authorization header not found", %{conn: conn} do
+      project = insert(:project)
+      Asset.seed_multiple_assets(project)
+      :timer.sleep(2500)
```
Comment (arjun289) [inline apps/acqdat_api/test/acqdat_api_web/controllers/elasticsearch/asset_controller_test.exs:117]: No need to seed here.

**Review by arjun289 — CHANGES_REQUESTED** [review]
- Please remove commented out code.
- Do not seed data when you are testing failure because of authorization related issues. It will only slow down the tests
- Also, seed in the setup block of `describe` so that you don't have to write it again and again in the tests.
- The retry for elastic search insert doesn't appear to be implemented properly.
- Seems while searching and indexing with elastic search, a DB query is being run to load the data for every entity found. Why don't you load all the data in just one query for all the entities found.
- While seeding multiple entities in elastic search instead of inserting just three entities take input from user and use Enum or comprehensions to iterate and seed.

File: apps/acqdat_api/lib/acqdat_api_web/controllers/iot-manager/gateway_controller.ex [hunk 4]
```diff
@@ -24,17 +25,61 @@ defmodule AcqdatApiWeb.IotManager.GatewayController do
...
+
+  #       conn
+  #       |> put_status(200)
+  #       |> render("index.json", gateway)
+
+  #     404 ->
+  #       conn
+  #       |> send_error(404, "Resource Not Found")
+  #   end
+  # end
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/iot-manager/gateway_controller.ex:66]: done

File: apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex [hunk 5]
```diff
@@ -24,24 +25,68 @@ defmodule AcqdatApiWeb.EntityManagement.ProjectController do
...
+  #       conn
+  #       |> put_status(200)
+  #       |> render("index.json", project)
+
+  #     404 ->
+  #       conn
+  #       |> send_error(404, "Resource Not Found")
+  #   end
+  # end
+
```
[the remark this answers is not in the document]
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex:89]: done

**Pull request merged**